packages/components/src/modal/index.tsx

```diff
@@ -106,7 +105,10 @@ function UnforwardedModal(
 		if ( openModalCount === 1 ) {
 			ariaHelper.hideApp( ref.current );
 			document.body.classList.add( bodyOpenClassName );
+		} else if ( openModalCount > 1 ) {
+			dismissOpenModal();
```

Comment:
> The only case useFocusOutside was actually supporting since https://github.com/WordPress/gutenberg/pull/52653 was the closing of one modal when another opened. So with the hook removed these changes maintain that behavior. I had missed this previously and we don’t have a test for it.

I tried to add code to test opening a modal inside a modal, and this PR instantly closes the new modal as described in the quoted text.

<details>
<summary>See code diff</summary>

```diff
diff --git a/packages/components/src/modal/stories/index.story.tsx b/packages/components/src/modal/stories/index.story.tsx
index 6963026963..6697d65346 100644
--- a/packages/components/src/modal/stories/index.story.tsx
+++ b/packages/components/src/modal/stories/index.story.tsx
@@ -48,11 +48,17 @@ export default meta;
 
 const Template: StoryFn< typeof Modal > = ( { onRequestClose, ...args } ) => {
 	const [ isOpen, setOpen ] = useState( false );
+	const [ isOpenTwo, setOpenTwo ] = useState( false );
 	const openModal = () => setOpen( true );
+	const openModalTwo = () => setOpenTwo( true );
 	const closeModal: ModalProps[ 'onRequestClose' ] = ( event ) => {
 		setOpen( false );
 		onRequestClose( event );
 	};
+	const closeModalTwo: ModalProps[ 'onRequestClose' ] = ( event ) => {
+		setOpenTwo( false );
+		onRequestClose( event );
+	};
 
 	return (
 		<>
@@ -86,9 +92,29 @@ const Template: StoryFn< typeof Modal > = ( { onRequestClose, ...args } ) => {
 						src="https://www.openstreetmap.org/export/embed.html?bbox=-0.004017949104309083%2C51.47612752641776%2C0.00030577182769775396%2C51.478569861898606&layer=mapnik"
 					/>
 
+					<Button variant="secondary" onClick={ openModalTwo }>
+						Open Modal two
+					</Button>
+
 					<Button variant="secondary" onClick={ closeModal }>
 						Close Modal
 					</Button>
+
+					{ isOpenTwo && (
+						<Modal
+							onRequestClose={ closeModalTwo }
+							style={ { maxWidth: '600px' } }
+							{ ...args }
+						>
+							Two!
+							<Button
+								variant="secondary"
+								onClick={ closeModalTwo }
+							>
+								Close Modal
+							</Button>
+						</Modal>
+					) }
 				</Modal>
 			) }
 		</>

```

</details>

But then I tried the same code diff on `trunk`, and I was able to open a nested modal (ie. neither the "old" modal, nor the new modal closed).

What is the correct expected behaviour? Is trunk currently experiencing a regression?

Reply:
The expected behavior is that only a single modal can be open at any time. I don't think trunk is experiencing a regression as I've not been able to reproduce it (testing in WP, by launching other modals with command palette). Perhaps there was a build hiccup when you were testing? I can test more soon though.

Reply:
I've tested more and trunk seems fine. I get one modal open at a time. Granted I don't see evidence that the behavior was by design it's just how it is (probably since #16878).

Reply:
As far as I've tested, only one modal is always displayed, both in trunk and in this PR. Probably there is a difference between trunk and this PR when Modal components are nested?

https://github.com/WordPress/gutenberg/assets/54422211/ec05dc2d-bd79-406c-ab22-a46d1be9e688

Reply:
My bad. I missed that your example was nested (the first diff wasn't applying for me). I've tested with your latest diff applied and see the same thing as you. I think the behavior makes sense yet I wondered since when has it been like that. I had my suspicions it's since #52653 and that seems to be the case. So I think the behavior would be worth preserving but it does appear to be a recent incidental "enhancement". 

<details><summary>A diff reverting Modal to before #52653</summary>

```diff
diff --git a/packages/components/src/modal/index.tsx b/packages/components/src/modal/index.tsx
index d21a3f9ae3..3a630eed11 100644
--- a/packages/components/src/modal/index.tsx
+++ b/packages/components/src/modal/index.tsx
@@ -171,34 +171,6 @@ function UnforwardedModal(
 		[ hasScrolledContent ]
 	);
 
-	let pressTarget: EventTarget | null = null;
-	const overlayPressHandlers: {
-		onPointerDown: React.PointerEventHandler< HTMLDivElement >;
-		onPointerUp: React.PointerEventHandler< HTMLDivElement >;
-	} = {
-		onPointerDown: ( event ) => {
-			if ( event.isPrimary && event.target === event.currentTarget ) {
-				pressTarget = event.target;
-				// Avoids loss of focus yet also leaves `useFocusOutside`
-				// practically useless with its only potential trigger being
-				// programmatic focus movement. TODO opt for either removing
-				// the hook or enhancing it such that this isn't needed.
-				event.preventDefault();
-			}
-		},
-		// Closes the modal with two exceptions. 1. Opening the context menu on
-		// the overlay. 2. Pressing on the overlay then dragging the pointer
-		// over the modal and releasing. Due to the modal being a child of the
-		// overlay, such a gesture is a `click` on the overlay and cannot be
-		// excepted by a `click` handler. Thus the tactic of handling
-		// `pointerup` and comparing its target to that of the `pointerdown`.
-		onPointerUp: ( { target, button } ) => {
-			const isSameTarget = target === pressTarget;
-			pressTarget = null;
-			if ( button === 0 && isSameTarget ) onRequestClose();
-		},
-	};
-
 	return createPortal(
 		// eslint-disable-next-line jsx-a11y/no-static-element-interactions
 		<div
@@ -208,7 +180,6 @@ function UnforwardedModal(
 				overlayClassName
 			) }
 			onKeyDown={ handleEscapeKeyDown }
-			{ ...( shouldCloseOnClickOutside ? overlayPressHandlers : {} ) }
 		>
 			<StyleProvider document={ document }>
 				<div

```
</details>

I've yet to determine how hard it would be to support this use case with the approach in this PR.

Reply:
I think it probably is the desired behavior regardless of whether there are existing use cases. Also, my comment above was a bit misleading as I now think Modal probably has supported nesting since its inception.

Reply:
I came to the same conclusion, ie. that **we should support nested modals**

> I've yet to determine how hard it would be to support this use case with the approach in this PR.

Let me know how that goes, hopefully it's not too complicated 🤞

Reply:
Does supporting nested modals mean closing the outer modal when the inner modal is opened?

I'm wondering if that's possible. This is because the inner modal is rendered within the outer modal, so if the outer modal disappears (is unmounted), the inner modal must also disappear as a result 🤔

Reply:
Yes, Aki, I believe you're right. There might be some way to visually hide the outer one but it certainly couldn't unmount or else the inner one goes too.

I did find a way to support the nesting that doesn't seem too complicated but I have to focus on some other stuff for a bit and I'll push it later.

Reply:
My expectation would be that, when the inner modal is opened, the outer modal stays open in the background.

Clicking on the backdrop / pressing ESC at this point would close the inner modal, but leave the outer modal visible.

> Does supporting nested modals mean closing the outer modal when the inner modal is opened?

Doing so would mean that two modals can't be open at a time, which basically would mean that nested modals are not really supported?

Reply:
I pushed some changes and now this supports nested modals more completely than trunk. I say more completely because while working on this I realized a nested modal doesn't accessibly hide the outer modal that launched it. Probably not a big deal because who is actually using nested modals? Still, technically I think it's the right thing to do.

Here’s a demo:

https://github.com/WordPress/gutenberg/assets/9000376/892b2398-c0de-436f-8476-4d60eb0e5c55

And of course…

<details><summary>A diff to add the story I tested with</summary>

```diff
diff --git a/packages/components/src/modal/stories/index.story.tsx b/packages/components/src/modal/stories/index.story.tsx
index 6963026963..be36e1fb6f 100644
--- a/packages/components/src/modal/stories/index.story.tsx
+++ b/packages/components/src/modal/stories/index.story.tsx
@@ -6,7 +6,7 @@ import type { StoryFn, Meta } from '@storybook/react';
 /**
  * WordPress dependencies
  */
-import { useState } from '@wordpress/element';
+import { createContext, useContext, useState } from '@wordpress/element';
 import { starEmpty, starFilled } from '@wordpress/icons';
 
 /**
@@ -127,3 +127,117 @@ WithHeaderActions.args = {
 WithHeaderActions.parameters = {
 	...Default.parameters,
 };
+
+const Cue: React.FC< {
+	children: React.ReactNode;
+	not?: boolean;
+	of: React.Dispatch< React.SetStateAction< boolean > >;
+} > = ( { children, not, of } ) => (
+	<Button variant="secondary" onClick={ () => of( ! not ) }>
+		{ children }
+	</Button>
+);
+
+const depth = createContext( 0 );
+
+const ModalWithDepth: React.FC< Parameters< typeof Template >[ 0 ] > = ( {
+	children,
+	...args
+} ) => {
+	const localDepth = useContext( depth );
+	return (
+		<depth.Provider value={ localDepth + 1 }>
+			<Modal
+				style={ { maxWidth: '600px' } }
+				{ ...args }
+				title={ `Depth ${ localDepth }: ${ args.title }` }
+			>
+				{ children }
+			</Modal>
+		</depth.Provider>
+	);
+};
+
+const Recursive: React.FC< Parameters< typeof Template >[ 0 ] > = ( args ) => {
+	const [ isOuterOpen, setOuterOpen ] = useState( false );
+	const [ isInnerOpen, setInnerOpen ] = useState( false );
+	const [ isAdjacentOpen, setAdjacentOpen ] = useState( false );
+
+	const localDepth = useContext( depth );
+
+	return (
+		<>
+			<Cue of={ setOuterOpen }>
+				{ localDepth === 0 ? 'Inception' : `🪆 ${ localDepth }` }
+			</Cue>
+			{ isOuterOpen && (
+				<ModalWithDepth
+					style={ { maxWidth: '600px' } }
+					{ ...args }
+					onRequestClose={ ( event ) => {
+						setInnerOpen( false );
+						setOuterOpen( false );
+						args.onRequestClose( event );
+					} }
+				>
+					<p>
+						Lorem ipsum dolor sit amet, consectetur adipiscing elit,
+						sed do eiusmod tempor incididunt ut labore et magna
+						aliqua. Ut enim ad minim veniam, quis nostrud
+						exercitation ullamco laboris nisi ut aliquip ex ea ea
+						commodo consequat. Duis aute irure dolor in
+						reprehenderit in voluptate velit esse cillum dolore eu
+						fugiat nulla pariatur. Excepteur sint occaecat cupidatat
+						non proident, sunt in culpa qui noticia deserunt mollit
+						anim id est laborum.
+					</p>
+
+					<InputControl style={ { marginBottom: '20px' } } />
+
+					<Cue of={ setInnerOpen }>🪆 { localDepth + 1 }</Cue>
+
+					<Cue of={ setAdjacentOpen }>➡️ Adjacent</Cue>
+
+					{ isInnerOpen && (
+						<ModalWithDepth
+							{ ...args }
+							onRequestClose={ () => {
+								setInnerOpen( false );
+								args.onRequestClose();
+							} }
+						>
+							<Recursive { ...args } />
+
+							<Cue of={ setAdjacentOpen }>
+								↖️ Adjacent to outer
+							</Cue>
+						</ModalWithDepth>
+					) }
+				</ModalWithDepth>
+			) }
+			{ isAdjacentOpen && (
+				<Modal
+					style={ { maxWidth: '500px' } }
+					{ ...args }
+					onRequestClose={ ( event ) => {
+						setAdjacentOpen( false );
+						args.onRequestClose( event );
+					} }
+					title={ `Depth ${ localDepth } adjacent: ${ args.title }` }
+				>
+					<p>
+						Lorem ipsum dolor sit amet, consectetur adipiscing elit,
+						sed do eiusmod tempor incididunt ut labore et magna
+						aliqua. Ut enim ad minim veniam, quis nostrud.
+					</p>
+				</Modal>
+			) }
+		</>
+	);
+};
+
+export const Recursion: StoryFn< typeof Modal > = ( args ) => (
+	<Recursive { ...args } />
+);
+Recursion.args = Default.args;
+Recursion.parameters = Default.parameters;

```
</details>

Reply:
Since #54569 and #54743 we have unit tests covering this so I'll mark this as resolved.